### PR TITLE
[ads] Implement AdsObserver

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/28726

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Confirm Reminders, brave://rewards, and upgrade browser required (426) are working as expected.

### CODE REVIEW

**Review by kylehickinson — APPROVED** [review]

File: browser/ui/webui/brave_rewards_page_ui.cc [hunk 1]
```diff
@@ -276,22 +280,25 @@ class RewardsDOMHandler
       const brave_rewards::RewardsNotificationService::RewardsNotificationsList&
           notifications_list) override;
 
-  // AdsServiceObserver implementation
-  void OnAdRewardsDidChange() override;
-  void OnNeedsBrowserUpgradeToServeAds() override;
+  // bat_ads::mojom::BatAdsObserver implementation
+  void OnBraveRewardsDidChange() override;
```
Comment (zenparsing) [inline browser/ui/webui/brave_rewards_page_ui.cc:284]: What event triggers this handler? Can we use a name that is a little more self-describing?

Comment (tmancey) [reply browser/ui/webui/brave_rewards_page_ui.cc:284]: This is triggered after successfully initializing ads on browser launch and when the statement of accounts changes for brave://rewards.

**Review by zenparsing — APPROVED** [review]
Rewards 👍

File: browser/ui/webui/brave_rewards_page_ui.cc (hunk 1 above)
Comment (tmancey) [reply browser/ui/webui/brave_rewards_page_ui.cc]: Fixed

File: browser/ui/webui/brave_rewards_page_ui.cc [hunk 2]
```diff
@@ -627,15 +634,15 @@ void RewardsDOMHandler::OnJavascriptAllowed() {
     rewards_service_observation_.Observe(rewards_service_);
   }
 
-  if (ads_service_) {
-    ads_service_observation_.Reset();
-    ads_service_observation_.Observe(ads_service_);
-  }
+  bat_ads_observer_receiver_.reset();
+  ads_service_->AddBatAdsObserver(
```
Comment (aseren) [inline browser/ui/webui/brave_rewards_page_ui.cc:638]: check condition if ads_service_ is not null was deleted

File: browser/ui/webui/new_tab_page/brave_new_tab_message_handler.cc [hunk 3]
```diff
@@ -312,15 +312,14 @@ void BraveNewTabMessageHandler::OnJavascriptAllowed() {
       base::BindRepeating(&BraveNewTabMessageHandler::OnPreferencesChanged,
                           base::Unretained(this)));
 
-  if (ads_service_) {
-    ads_service_observation_.Reset();
-    ads_service_observation_.Observe(ads_service_);
-  }
+  bat_ads_observer_receiver_.reset();
+  ads_service_->AddBatAdsObserver(
```
Comment (aseren) [inline browser/ui/webui/new_tab_page/brave_new_tab_message_handler.cc:316]: `  if (ads_service_) {` is missing

Comment (tmancey) [reply browser/ui/webui/new_tab_page/brave_new_tab_message_handler.cc]: Fixed

File: browser/ui/webui/brave_rewards_page_ui.cc (hunk 2 above)
Comment (tmancey) [reply browser/ui/webui/brave_rewards_page_ui.cc]: Fixed

**Review by aseren — APPROVED** [review]

File: components/brave_ads/core/internal/ads_notifier_manager.cc [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+AdsNotifierManager::~AdsNotifierManager() = default;
+
+// static
+AdsNotifierManager& AdsNotifierManager::GetInstance() {
+  return GlobalState::GetInstance()->GetAdsNotifierManager();
+}
+
+void AdsNotifierManager::AddObserver(
+    std::unique_ptr<AdsObserverInterface> observer) {
+  observers_.push_back(std::move(observer));
```
Comment (aseren) [inline components/brave_ads/core/internal/ads_notifier_manager.cc:26]: CHECK(observer);

Comment (tmancey) [reply components/brave_ads/core/internal/ads_notifier_manager.cc:26]: Thanks, will add in a followup.

**Pull request merged**